File: typescript/resistor-color-trio/resistor-color-trio.ts
const COLORS = [
	`black`,
	`brown`,
	`red`,
	`orange`,
	`yellow`,
	`green`,
	`blue`,
	`violet`,
	`grey`,
	`white`,
] as const;

export type Color = (typeof COLORS)[number];

const ohms = [
	[1_000_000_000, "giga"],
	[1_000_000, "mega"],
	[1_000, "kilo"],
] as const;

export function decodedResistorValue([band1, band2, band3]: Color[]): string {

	const power = 10 ** COLORS.indexOf(band3);
  const num = ((COLORS.indexOf(band1) * 10) + COLORS.indexOf(band2)) * (power)
	const [divisor, prefix] = ohms.find(([divisor]) => num >= divisor) ?? [1, ""];
	return `${num / divisor} ${prefix}ohms`;
}

File: solutions/python/resistor-color-trio/1/resistor_color_trio.py
def label(colors):
    color_map = {
        "black": 0,
        "brown": 1,
        "red": 2,
        "orange": 3,
        "yellow": 4,
        "green": 5,
        "blue": 6,
        "violet": 7,
        "grey": 8,
        "white": 9,
    }

    first, second, multiplier = colors[:3]
    value = (color_map[first] * 10 + color_map[second]) * (10 ** color_map[multiplier])

    return _format_label(value)


def _format_label(value):
    if value >= 1_000_000_000:
        return f"{_format_value(value, 1_000_000_000)} gigaohms"
    if value >= 1_000_000:
        return f"{_format_value(value, 1_000_000)} megaohms"
    if value >= 1_000:
        return f"{_format_value(value, 1_000)} kiloohms"
    return f"{value} ohms"


def _format_value(value, unit):
    number = value // unit
    if value % unit == 0:
        return str(number)

    whole, remainder = divmod(value, unit)
    remainder_str = str(remainder).rjust(len(str(unit)) - 1, "0").rstrip("0")
    return f"{whole}.{remainder_str}"
